crates/biome_js_analyze/src/lint/nursery/no_label_without_control.rs

```diff
@@ -0,0 +1,281 @@
+    /// <label for="js_id" aria-labelledby="A label" />;
+    /// <label htmlFor="js_id" aria-label="A label" />;
+    /// <label htmlFor="js_id" aria-labelledby="A label" />;
+    /// <label>A label<input /></label>;
+    /// <label>A label<textarea /></label>;
+    /// <label><img alt="A label" /><input /></label>;
+    /// ```
+    ///
+    /// ## Options
+    ///
+    /// The rule supports the following options:
+    /// - `inputComponents` - An array of component names that should be considered the same as an `input` element.
+    /// - `labelAttributes` - An array of attributes that should be treated as the `label` accessible text content.
+    /// - `labelComponents` - An array of component names that should be considered the same as a `label` element.
+    ///
+    /// Both options `inputComponents` and `labelComponents` don't have support for namespace components (e.g. `<Control.Input>`).
+    ///
+    /// ```json
+    /// {
+    ///     "//": "...",
+    ///     "options": {
+    ///         "inputComponents": ["CustomInput"],
+    ///         "labelAttributes": ["label"],
+    ///         "labelComponents": ["CustomLabel"]
+    ///     }
+    /// }
+    /// ```
+    ///
+    pub NoLabelWithoutControl {
+        version: "next",
+        name: "noLabelWithoutControl",
+        language: "jsx",
+        sources: &[RuleSource::EslintJsxA11y("label-has-associated-control")],
+        recommended: true,
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Deserializable, Eq, PartialEq, Serialize)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoLabelWithoutControlOptions {
+    /// Array of component names that should be considered the same as an `input` element.
+    pub input_components: Vec<String>,
+    /// Array of attributes that should be treated as the `label` accessible text content.
+    pub label_attributes: Vec<String>,
+    /// Array of component names that should be considered the same as a `label` element.
+    pub label_components: Vec<String>,
+}
+
+pub struct NoLabelWithoutControlState {
+    pub has_text_content: bool,
+    pub has_control_association: bool,
+}
+
+impl Rule for NoLabelWithoutControl {
+    type Query = Ast<AnyJsxTag>;
+    type State = NoLabelWithoutControlState;
+    type Signals = Option<Self::State>;
+    type Options = NoLabelWithoutControlOptions;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let options = ctx.options();
+        let label_attributes = &options.label_attributes;
+        let label_components = &options.label_components;
+        let input_components = &options.input_components;
+        let default_label_attributes = &["aria-label", "alt"];
+        let default_label_components = &["label"];
+        let default_input_components =
+            &["input", "meter", "output", "progress", "select", "textarea"];
```

Comment:
You're creating the same array every time you run the rule. This isn't very efficient. Create a static array instead, for example:

```rs
const  DEFAULT_LABEL_ATTIRBUTES: [&str; 2] = &["aria-label", "alt"];
```

You'll have to place at the top level of the file, for example

https://github.com/biomejs/biome/blob/8a26e257636ecf24e2543ee0f1af3f957ffbbf5c/crates/biome_js_analyze/src/globals/javascript/node.rs#L4

Reply:
good catch, done!